### PR TITLE
new_audit: lcp-lazy-loaded

### LINKED ISSUE #12785: New audit: LCP element was lazy-loaded
**Feature request summary**

I'd like to see a new performance or best practices audit that fails if the LCP element was lazy-loaded.

**What is the motivation or use case for changing this?**
Lazy-loading images above the fold may adversely affect LCP performance.

**How is this beneficial to Lighthouse?**
It identifies an opportunity to improve the performance of LCP, a Core Web Vital metric.

### PR DESCRIPTION
**Pull request by milutin**
Fixes #12785

### CODE REVIEW

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:2]:
```suggestion
 * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
```

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+class LargestContentfulPaintElement extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'lcp-lazy-loaded',
+      title: str_(UIStrings.title),
+      description: str_(UIStrings.description),
+      scoreDisplayMode: Audit.SCORING_MODES.ERROR,
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:29]:
```suggestion
```

binary is the default and that's what we want here :)

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      supportedModes: ['navigation'],
+      requiredArtifacts: ['traces', 'TraceElements', 'ImageElements'],
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:31]: I don't think we actually use `traces` anywhere
```suggestion
      requiredArtifacts: ['TraceElements', 'ImageElements'],
```

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    ];
+
+    const details = Audit.makeTableDetails(headings, lcpElementDetails);
+
+    const displayValue = str_(i18n.UIStrings.displayValueElementsFound,
+      {nodeCount: lcpElementDetails.length});
+
+    return {
+      score: 1,
+      notApplicable: lcpElementDetails.length === 0,
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:70]: there are some good edge cases discussed in https://github.com/GoogleChrome/lighthouse/issues/12785 , we probably also want to mark this as notApplicable if the scroll position isn't at the top (in case we measured a URL with a fragment and out-of-viewport anchor)

I don't believe we currently capture this information anywhere in our artifacts, but I think it makes a lot of sense in `ViewportDimensions`, we might discuss in eng sync next week :)

FWIW, our `notApplicable` logic can also be a little confusing so we tend to have an explicit branch for it

```js
if (isAuditNotApplicable()) {
  return {score: 1, notApplicable: true};
}

return {
  score: // the actual results...
}

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      {key: 'node', itemType: 'node', text: str_(i18n.UIStrings.columnElement)},
+    ];
+
+    const details = Audit.makeTableDetails(headings, lcpElementDetails);
+
+    const displayValue = str_(i18n.UIStrings.displayValueElementsFound,
+      {nodeCount: lcpElementDetails.length});
+
+    return {
+      score: 1,
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:69]:
```suggestion
      score: 0,
```

File: lighthouse-core/config/default-config.js [hunk 6]
```diff
@@ -452,6 +453,7 @@ const defaultConfig = {
         {id: 'duplicated-javascript', weight: 0, group: 'load-opportunities'},
         {id: 'legacy-javascript', weight: 0, group: 'load-opportunities'},
         {id: 'preload-lcp-image', weight: 0, group: 'load-opportunities'},
+        {id: 'lcp-lazy-loaded', weight: 0, group: 'load-opportunities'},
```
Comment (patrickhulce) [inline lighthouse-core/config/default-config.js:456]: we only put an audit into `load-opportunities` today if we can quantify the impact on resources, for now let's leave it as `diagnostics`

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    /** @type {LH.Audit.Details.Table['headings']} */
+    const headings = [
+      {key: 'node', itemType: 'node', text: str_(i18n.UIStrings.columnElement)},
+    ];
+
+    const details = Audit.makeTableDetails(headings, lcpElementDetails);
+
+    const displayValue = str_(i18n.UIStrings.displayValueElementsFound,
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:65]: there can only ever be 1 here, so I don't know if it's worth having a display value

**Review by patrickhulce — CHANGES_REQUESTED** [review]
thanks @milutin! you pieced together the right info, nicely done :) 

we'll probably discuss the exact policy a bit more in our eng sync, so a few more comments might be coming

File: lighthouse-core/audits/lcp-lazy-loaded.js (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-core/audits/lcp-lazy-loaded.js:66]: alright we discussed this eng sync and we're fine to proceed here, we just need to account for the not applicable cases:

- gather mode is not `navigation` (you're already doing this 👍 )
- the viewport top is greater than 1/4 of the viewportHeight

Comment (milutin) [reply lighthouse-core/audits/lcp-lazy-loaded.js:66]: Great, thanks. I would like to clarify what do you mean viewport top and why greater than 1/4? Would something like this work for you? 
```
        const imageTop = lcpImageElement.clientRect.top;
        const viewportHeight = artifacts.ViewportDimensions.innerHeight;
        if (imageTop / viewportHeight * 100 > 25) {
          lcpElementDetails.push({
            node: Audit.makeNodeItem(lcpImageElement.node),
          });
        }
```

Comment (patrickhulce) [reply lighthouse-core/audits/lcp-lazy-loaded.js:66]: Yes that's another way to go about it! Though I think we want a simple `imageTop < viewportHeight` if we're approaching it from the image clientRect angle.

The goal here is the ensure we're marking the audit notApplicable when we're analyzing anything other than the top of the page (LCP element being in the first viewport *technically* could still be wrong if it rendered *before* the scroll, but seems like we can live with that edge case for now)

**Comment by milutin** [comment]
Sorry for break. I am ready to finish this. I am working on tests to pass checks and on strings.

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const i18n = require('../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Descriptive title of a diagnostic audit that provides */
+  title: 'Largest Contentful Paint element was lazy-loaded',
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:13]: Binary scored audits should have two versions of the title, one that expresses when it is passing and one that expresses when it is failing :)

```suggestion
  /** Title of a Lighthouse audit that provides detail on whether the largest above-the-fold image was loaded with sufficient priority. This descriptive title is shown to users when the image was loaded properly. */
  title: 'Largest Contentful Paint image was not lazily loaded',
    /** Title of a Lighthouse audit that provides detail on whether the largest above-the-fold image was loaded with sufficient priority. This descriptive title is shown to users when the image was loaded inefficiently using the `loading=lazy` attribute. */
  failureTitle: 'Largest Contentful Paint image was lazily loaded',
```

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  /** Description of a Lighthouse audit that tells */
+  description: 'Consider to remove lazy loading for largest contentful paint element.',
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:15]:
```suggestion
  /** Description of a Lighthouse audit that tells the user why the advice is important. This is displayed after a user expands the section to see more. No character length limits. */
  description: 'Above-the-fold images that are lazily loaded render later in the page lifecycle, which can delay the largest contentful paint. [Learn more](https://web.dev/lcp-lazy-loading/).',
```

perhaps?

File: lighthouse-core/audits/lcp-lazy-loaded.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    return imageTop < viewportHeight;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {LH.Audit.Product}
+   */
+  static audit(artifacts) {
+    const images = artifacts.ImageElements;
+    const lazyLoadedImages = images.filter(
```
Comment (patrickhulce) [inline lighthouse-core/audits/lcp-lazy-loaded.js:51]: We can be a little more specific with our audit results if we avoid filtering to lazy from the start.

something more like

```js
const lcpElmentImage = images.find(elem => elem.node.devtoolsNodePath === lcpElement.node.devtoolsNodePath);

if (!lcpElementImage || !this.isImageInViewport(lcpElementImage, artifacts.ViewportDimensions)) {
  return {score: 1, notApplicable: true};
}

const details = /* as-is ... */ {};
return {
  score: lcpElementImage.loading === 'lazy' ? 0 : 1, 
  details,
}
```

This will provide clearer not applicable and passing states :)

File: lighthouse-core/test/audits/lcp-lazy-loaded-test.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const LargestContentfulPaintLazyLoaded =
+  require('../../audits/lcp-lazy-loaded.js');
+
+/* eslint-env jest */
+describe('Performance: lcp-lazy-loaded audit', () => {
+  it('correctly surfaces the lazy loaded LCP element', async () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/lcp-lazy-loaded-test.js:13]: could we get coverage of the viewport notapplicable case and the passing case as well?

**Review by patrickhulce — CHANGES_REQUESTED** [review]
taking great shape @milutin thank you! :) 

this would also be a great candidate for a smoketest if you'd like to add loading lazy to https://github.com/GoogleChrome/lighthouse/blob/47c2bc5d96d87f121a0ada42ddf60090f98a6339/lighthouse-cli/test/fixtures/perf/delayed-element.js#L17-L19 and the audit expectations to https://github.com/GoogleChrome/lighthouse/blob/47c2bc5d96d87f121a0ada42ddf60090f98a6339/lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js#L283-L297 :)

**Comment by milutin** [comment]
I was looking on smoke tests, it will take me awhile to add one.
When I added lazy loading in delayed-element
```
imgEl.src = '../dobetterweb/lighthouse-480x318.jpg';
imgEl.loading = 'lazy';
```

Smoke tests failed with difference at TraceElement artifacts. 

I will look later how to change expected traceElement artifacts to pass smoke test and how to add expectations of failing lcp-lazy-loaded audit.

**Review by patrickhulce — APPROVED** [review]
excellent job @milutin!

**Comment by milutin** [comment]
Great, thanks a lot for help and guidance!

**Comment by sethta** [comment]
We have been getting conflicting results with this specific check. On every site we have tested using the [Trellis WordPress theme](https://www.mediavine.com/products/trellis/), we have found that _not_ lazy loading the LCP has caused our scores to go down, both on mobile and desktop. We typically see a 50% to 200% increase in performance for the LCP time when we lazy load that image.

This makes it difficult when our publishers get the "Largest Contentful Paint image was lazily loaded" diagnostic, with the large red error triangle next to it, but we have not found a single example of this recommendation speeding up the site. Even the [Google Developers documentation](https://web.dev/browser-level-image-lazy-loading/#avoid-lazy-loading-images-that-are-in-the-first-visible-viewport) states the following:

> In Chromium, the impact of images in the initial viewport being marked with `loading=lazy` on Largest Contentful Paint is fairly small, with a regression of <1% at the 75th and 99th percentiles compared to eagerly loaded images.

Our interpretation of this is that there is too much emphasis placed on this audit. The potential regression is extremely minimal, and with every site we have tested this showing the exact opposite, we are worried that this perceived requirement will cause more harm than good for website owners using Trellis.

**Comment by rviscomi** [comment]
@sethta what tool are you using to test and are you able to share the results?

Also note that we reassessed the performance impact of lazy-loading the LCP image and found clear evidence that it's harmful: https://web.dev/lcp-lazy-loading/

**Comment by sethta** [comment]
@rviscomi

We are typically using the Lighthouse audit in the Chrome dev tools, but because most of our publishers on Trellis don't know how to access that, we also use PageSpeed Insights.

I've replicated one of our testing sites so it's publicly available, and the results still show that lazy-loading the LCP image is faster on sites using Trellis.

These two pages are completely identical with the exception that the primary image is lazy loaded:
- [LCP Lazy Loaded](https://trellis.alling.dev/lcp-lazy-loaded/)
- [LCP Not Lazy Loaded](https://trellis.alling.dev/lcp-not-lazy-loaded/)

### Lazy Loaded
https://pagespeed.web.dev/report?url=https%3A%2F%2Ftrellis.alling.dev%2Flcp-lazy-loaded%2F

When we run the tests on the page with the lazy loaded image, we consistently get Performance scores of 100 for both Mobile and Desktop results.

**Mobile GPSI: 100**
image

**Other Images:**
- [Mobile Lighthouse Audit: 100](https://user-images.githubusercontent.com/220730/163034579-8be93b2b-908c-409f-ba4c-e3a2b420e4a0.png)
- [Desktop GPSI: 100](https://user-images.githubusercontent.com/220730/163034026-c629cbef-0e3f-4f29-9404-a79d854f4903.png)
- [Desktop Lighthouse Audit: 100](https://user-images.githubusercontent.com/220730/163034826-e813805b-fe01-475e-8484-876cbe057b2e.png)

### NOT Lazy Loaded
https://pagespeed.web.dev/report?url=https%3A%2F%2Ftrellis.alling.dev%2Flcp-not-lazy-loaded%2F

When we run the tests on the page where we have removed lazy loading, we have never been able to achieve a Performance higher than 96, directly related to slow LCP, for Mobile and the LCP time was consistently slower on Desktop (0.5-0.8s compared to 0.3s).

**Mobile GPSI: 94**
image

**Other Images:**
- [Mobile Lighthouse Audit: 96](https://user-images.githubusercontent.com/220730/163036230-2cb61d6c-ef7f-466a-bc90-717693c0bb7e.png)
- [Desktop GPSI: 100](https://user-images.githubusercontent.com/220730/163036496-946625cb-a8b8-4b27-82df-2d1034d09344.png)
- [Desktop Lighthouse Audit: 100](https://user-images.githubusercontent.com/220730/163036028-7ff38eb0-9234-493a-84a1-92f14a22d036.png)

**Comment by rviscomi** [comment]
Thanks for the additional info @sethta.

It looks to me like your server might be too inconsistent for these results to be meaningful.

Here's a [test](https://www.webpagetest.org/graph_page_data.php?tests=220412_AiDcM9_GBK&medianMetric=chromeUserTiming.LargestContentfulPaint&fv=1&median_run=1&control=NOSTAT#TTFB) of 9 runs of the lcp-not-lazy-loaded page and you can see that the backend time to first byte (TTFB) metric varies between about 500ms to 4000ms. That will greatly affect the LCP time, much more so than lazy loading.

image

Interestingly, the [test](https://www.webpagetest.org/graph_page_data.php?tests=220412_AiDcC5_GBP&medianMetric=chromeUserTiming.LargestContentfulPaint&fv=1&median_run=1&control=NOSTAT#TTFB) of the lcp-lazy-loaded page shows that TTFB is more consistent. Is it possible that you're doing some kind of slow rewrite on the server to generate the not-lazy-loaded page?

image

Even when we pick tests with similar TTFB times to [compare](https://www.webpagetest.org/video/compare.php?tests=220412_AiDcC5_GBP-r%3A3%2C220412_AiDcM9_GBK-r%3A2&highlightLCP=1&thumbSize=200&ival=100&end=visual), the eager version still appears to render the image more quickly:

image

The lazy page has a TTFB of 572ms and the eager page has a TTFB of 620ms. However, the lazy page is reported to have an LCP of [653ms](https://www.webpagetest.org/vitals.php?test=220412_AiDcC5_GBP&run=3&cached=0#lcp) while the eager page has an LCP of [1114ms](https://www.webpagetest.org/vitals.php?test=220412_AiDcM9_GBK&run=2&cached=0#lcp). What's happening is that your lazy version is actually fooling the LCP calculation by drawing a gray box where the image should be:

```html
src="data:image/svg+xml,%3Csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%20683%201024'%3E%3Crect%20width='683'%20height='1024'%20style='fill:%23e3e3e3'/%3E%3C/svg%3E"
```

Unbeknownst to the browser, it thinks this inline SVG image is "contentful" so it considers this very early paint time as the LCP. But to a real user, the image doesn't appear loaded until about **1300ms**. This is really bad because you're optimizing for something that actually makes the UX worse, despite the metrics seeming to improve.

I'd strongly recommend removing the lazy loading behavior from your LCP images. These lab tests confirm that the image will actually render more quickly by loading the images eagerly. Even if your measured LCP seems to get slightly worse, you're making the UX better, which is what matters most. You should also assume that at some point browsers will patch the buggy behavior, so the sooner you move away from relying on it the better.

**Comment by sethta** [comment]
@rviscomi Thank you so much for this!

We have been banging our heads against the wall for a couple of weeks trying to figure out why we were getting different results.

The SVG application was developed a few years ago as a way to handle lazy-loaded images through a Javascript solution before the native browser solution existed, and it has run as a fallback ever since, as it has helped with both speeds and CLS.

One of our plans is to remove that application now that Safari supports lazy loaded images, so I think you just saved us from a real headache when we get that feature adjusted.

Thanks again!
